stdlib/universe/mode.go

```diff
@@ -0,0 +1,584 @@
+			}
+			storedVals := []uint64{}
+			n := int64(0)
+			for k := range uintMode {
+				if uintMode[k] > n {
+					storedVals = storedVals[:0]
+					storedVals = append(storedVals, k)
+					n = uintMode[k]
+				} else if uintMode[k] == n {
+					storedVals = append(storedVals, k)
+				}
+			}
+			if numNil > n {
+				if err := builder.AppendNil(colIdx); err != nil {
+					return err
+				}
+			} else if len(storedVals) == len(uintMode) && (numNil == 0 || numNil == n) {
+				if err := builder.AppendNil(colIdx); err != nil {
+					return err
+				}
+			} else if numNil == n {
+				sort.Slice(storedVals, func(i, j int) bool { return storedVals[i] < storedVals[j] })
+				for j := range storedVals {
+					if err := builder.AppendUInt(colIdx, storedVals[j]); err != nil {
+						return err
+					}
+				}
+				if err := builder.AppendNil(colIdx); err != nil {
+					return err
+				}
+			} else {
+				sort.Slice(storedVals, func(i, j int) bool { return storedVals[i] < storedVals[j] })
+				for j := range storedVals {
+					if err := builder.AppendUInt(colIdx, storedVals[j]); err != nil {
+						return err
+					}
+				}
+			}
+			if err := execute.AppendKeyValues(tbl.Key(), builder); err != nil {
+				return err
+			}
+
+			return nil
+		})
+	case flux.TFloat:
+		floatMode = make(map[float64]int64)
+		return tbl.Do(func(cr flux.ColReader) error {
+			l := cr.Len()
+			for i := 0; i < l; i++ {
+				if cr.Floats(j).IsNull(i) {
+					numNil++
+					continue
+				}
+				v := cr.Floats(j).Value(i)
+				floatMode[v]++
+			}
+			storedVals := []float64{}
+			n := int64(0)
+			for k := range floatMode {
+				if floatMode[k] > n {
+					storedVals = storedVals[:0]
+					storedVals = append(storedVals, k)
+					n = floatMode[k]
+				} else if floatMode[k] == n {
+					storedVals = append(storedVals, k)
+				}
+			}
+			if numNil > n {
+				if err := builder.AppendNil(colIdx); err != nil {
+					return err
+				}
+			} else if len(storedVals) == len(floatMode) && (numNil == 0 || numNil == n) {
+				if err := builder.AppendNil(colIdx); err != nil {
+					return err
+				}
+			} else if numNil == n {
+				sort.Float64s(storedVals)
+				for j := range storedVals {
+					if err := builder.AppendFloat(colIdx, storedVals[j]); err != nil {
+						return err
+					}
+				}
+				if err := builder.AppendNil(colIdx); err != nil {
+					return err
+				}
+			} else {
+				sort.Float64s(storedVals)
+				for j := range storedVals {
+					if err := builder.AppendFloat(colIdx, storedVals[j]); err != nil {
+						return err
+					}
+				}
+			}
+			if err := execute.AppendKeyValues(tbl.Key(), builder); err != nil {
+				return err
+			}
+
+			return nil
+		})
+	case flux.TString:
+		stringMode = make(map[string]int64)
+		return tbl.Do(func(cr flux.ColReader) error {
+			l := cr.Len()
+			for i := 0; i < l; i++ {
+				if cr.Strings(j).IsNull(i) {
+					numNil++
+					continue
+				}
+				v := cr.Strings(j).ValueString(i)
+				stringMode[v]++
+			}
+			storedVals := []string{}
```

Comment:
I think we can simplify the logic below since it seems to be fairly complex. Instead of playing weird tricks with a byte slice, we can feasibly just iterate over the map twice. The first time we iterate we are looking for the number of times a value has been seen and the number of values that match that frequency and the second time we construct the stored values. It would also be possible to optimize that second pass in certain circumstances.

The logical steps:

1. Iterate over the mode frequencies.
2. If a key has been seen more frequently than another key, then store the new key.
3. If a key has been seen with the same frequency as another key, increment the number of total values.

Here is some pseudocode.

```go
curr, n, total := "", int64(0)
for k := range stringMode {
    if stringMode[k] > n {
        curr, n, total = k, stringMode[k], 1
    } else if stringMode[k] == n {
        total++
    }
}

storedVals := make([]string, 0, len(total))
for k := range stringMode {
    if stringMode[k] == n {
        storedVals = append(storedVals, k)
    }
}
```

The variable names are bad so it might be a good idea to rename those.